django/db/models/fields/tuple_lookups.py

```diff
@@ -0,0 +1,222 @@
+import itertools
+
+from django.core.exceptions import EmptyResultSet
+from django.db.models.expressions import Cols
+from django.db.models.lookups import (
+    Exact,
+    GreaterThan,
+    GreaterThanOrEqual,
+    In,
+    IsNull,
+    LessThan,
+    LessThanOrEqual,
+)
+
+
+class TupleLookupMixin:
+    def get_prep_lookup(self):
+        self.check_tuple_lookup()
+        return super().get_prep_lookup()
+
+    def check_tuple_lookup(self):
+        assert isinstance(self.lhs, Cols)
+        self.check_rhs_is_tuple_or_list()
+        self.check_rhs_length_equals_lhs_length()
+
+    def check_rhs_is_tuple_or_list(self):
+        if not isinstance(self.rhs, (tuple, list)):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                "must be a tuple or a list"
+            )
+
+    def check_rhs_length_equals_lhs_length(self):
+        if len(self.lhs) != len(self.rhs):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                f"must have {len(self.lhs)} elements"
+            )
+
+    def check_rhs_is_collection_of_tuples_or_lists(self):
+        if not all(isinstance(vals, (tuple, list)) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                f"must be a collection of tuples or lists"
+            )
+
+    def check_rhs_elements_length_equals_lhs_length(self):
+        if not all(len(self.lhs) == len(vals) for vals in self.rhs):
+            raise ValueError(
+                f"'{self.lookup_name}' lookup of '{self.lhs.field.name}' field "
+                f"must have {len(self.lhs)} elements each"
+            )
+
+
+class TupleExact(TupleLookupMixin, Exact):
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, WhereNode
+
+        # e.g.: (a, b, c) == (x, y, z) as SQL:
```

Comment:
The `compiler.compile` usage would allow `as_sqlite` and `as_postgres` to be eventually implemented to use

```python
class Tuple(Func):
    function = ""

def as_sql(self, compile, connection):
    return compiler.compile(Exact(Tuple(*cols), Tuple(*self.rhs)))
```

Actually it seems that `Exact(Tuple(*cols), Tuple(*self.rhs))` works on all supported backends.

Reply:
Alright, this came up multiple times so let me look into it.
Django can be used with other backends too though, so maybe we should add feature flags for this? `supports_tuple_exact`, `supports_tuple_in`, etc.

Reply:
I went with the `as_<vendor>` functions (see the main branch), I was just wondering about the feature flags because of the 3rd party backends (as they cannot easily add `as_<vendor>`).